File: packages/tester/harmony/react_native_openharmony/src/main/cpp/RNOH/Performance/RNOHMarker.cpp
/**
 * Copyright (c) 2024 Huawei Technologies Co., Ltd.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "RNOHMarker.h"
#include <glog/logging.h>
#include "RNOH/Assert.h"

namespace rnoh {

std::vector<RNOHMarker::RNOHMarkerListener*> RNOHMarker::listeners = {};

void RNOHMarker::setLogMarkerIfNeeded() {
  static std::once_flag flag{};
  std::call_once(flag, []() {
    ReactMarker::logTaggedMarkerImpl = RNOHMarker::logMarker;
    ReactMarker::logTaggedMarkerBridgelessImpl = RNOHMarker::logMarker;
  });
}

void RNOHMarker::logMarker(RNOHMarkerId markerId) {
  logMarker(markerId, "");
}

void RNOHMarker::logMarker(RNOHMarkerId markerId, facebook::react::Tag tag) {
  logMarker(markerId, std::to_string(tag).c_str());
}

void RNOHMarker::logMarker(RNOHMarkerId markerId, const char* tag) {
  logMarker(markerId, tag, facebook::react::JSExecutor::performanceNow());
}

void RNOHMarker::logMarker(const std::string& markerId, const char* tag) {
  logMarker(harmonyMarkerIdForString(markerId), tag);
}

void RNOHMarker::logMarker(
    ReactMarker::ReactMarkerId markerId,
    const char* tag) {
  logMarker(harmonyMarkerIdForReactMarkerId(markerId), tag);
}

void RNOHMarker::logMarker(
    RNOHMarkerId markerId,
    const char* tag,
    double timestamp) {
  auto tagStr = tag == nullptr ? "" : std::string(tag);
  for (auto& listener : listeners) {
    listener->logMarker(markerId, tagStr, timestamp);
  }
}

void RNOHMarker::setAppStartTime(double appStartTime) {
  sAppStartTime = appStartTime;
}

double RNOHMarker::getAppStartTime() {
  return sAppStartTime;
}

void RNOHMarker::addListener(RNOHMarkerListener* listener) {
  listeners.push_back(listener);
}

void RNOHMarker::removeListener(RNOHMarkerListener* listener) {
  listeners.erase(
      std::remove(listeners.begin(), listeners.end(), listener),
      listeners.end());
}

RNOHMarker::RNOHMarkerId RNOHMarker::harmonyMarkerIdForReactMarkerId(
    const ReactMarker::ReactMarkerId markerId) {
  switch (markerId) {
    case ReactMarker::ReactMarkerId::RUN_JS_BUNDLE_START:
      return RNOHMarkerId::RUN_JS_BUNDLE_START;
    case ReactMarker::ReactMarkerId::RUN_JS_BUNDLE_STOP:
      return RNOHMarkerId::RUN_JS_BUNDLE_STOP;
    case ReactMarker::ReactMarkerId::CREATE_REACT_CONTEXT_STOP:
      return RNOHMarkerId::CREATE_REACT_CONTEXT_STOP;
    case ReactMarker::ReactMarkerId::JS_BUNDLE_STRING_CONVERT_START:
      return RNOHMarkerId::JS_BUNDLE_STRING_CONVERT_START;
    case ReactMarker::ReactMarkerId::JS_BUNDLE_STRING_CONVERT_STOP:
      return RNOHMarkerId::JS_BUNDLE_STRING_CONVERT_STOP;
    case ReactMarker::ReactMarkerId::NATIVE_MODULE_SETUP_START:
      return RNOHMarkerId::NATIVE_MODULE_SETUP_START;
    case ReactMarker::ReactMarkerId::NATIVE_MODULE_SETUP_STOP:
      return RNOHMarkerId::NATIVE_MODULE_SETUP_STOP;
    case ReactMarker::ReactMarkerId::REGISTER_JS_SEGMENT_START:
      return RNOHMarkerId::REGISTER_JS_SEGMENT_START;
    case ReactMarker::ReactMarkerId::REGISTER_JS_SEGMENT_STOP:
      return RNOHMarkerId::REGISTER_JS_SEGMENT_STOP;
    case ReactMarker::ReactMarkerId::NATIVE_REQUIRE_START:
      return RNOHMarkerId::NATIVE_REQUIRE_START;
    case ReactMarker::ReactMarkerId::NATIVE_REQUIRE_STOP:
      return RNOHMarkerId::NATIVE_REQUIRE_STOP;
    case ReactMarker::ReactMarkerId::REACT_INSTANCE_INIT_START:
      return RNOHMarkerId::REACT_INSTANCE_INIT_START;
    case ReactMarker::ReactMarkerId::REACT_INSTANCE_INIT_STOP:
      return RNOHMarkerId::REACT_INSTANCE_INIT_STOP;
    case ReactMarker::ReactMarkerId::APP_STARTUP_START:
      return RNOHMarkerId::APP_STARTUP_START;
    case ReactMarker::ReactMarkerId::APP_STARTUP_STOP:
      return RNOHMarkerId::APP_STARTUP_STOP;
    case ReactMarker::ReactMarkerId::INIT_REACT_RUNTIME_START:
      return RNOHMarkerId::INIT_REACT_RUNTIME_START;
    case ReactMarker::ReactMarkerId::INIT_REACT_RUNTIME_STOP:
      return RNOHMarkerId::INIT_REACT_RUNTIME_STOP;
    default:
      DLOG(WARNING) << "Unhandled ReactMarkerId " << markerId;
      RNOH_ASSERT(false);
      return RNOHMarkerId::UNKNOWN_REACT_MARKER;
  }
}

std::string RNOHMarker::harmonyMarkerIdToString(const RNOHMarkerId markerId) {
  switch (markerId) {
    case RNOHMarkerId::UNKNOWN_REACT_MARKER:
      return "UNKNOWN_REACT_MARKER";
    case RNOHMarkerId::APP_STARTUP_START:
      return "APP_STARTUP_START";
    case RNOHMarkerId::APP_STARTUP_STOP:
      return "APP_STARTUP_STOP";
    case RNOHMarkerId::INIT_REACT_RUNTIME_START:
      return "INIT_REACT_RUNTIME_START";
    case RNOHMarkerId::INIT_REACT_RUNTIME_STOP:
      return "INIT_REACT_RUNTIME_STOP";
    case RNOHMarkerId::NATIVE_REQUIRE_START:
      return "NATIVE_REQUIRE_START";
    case RNOHMarkerId::NATIVE_REQUIRE_STOP:
      return "NATIVE_REQUIRE_STOP";
    case RNOHMarkerId::RUN_JS_BUNDLE_START:
      return "RUN_JS_BUNDLE_START";
    case RNOHMarkerId::RUN_JS_BUNDLE_STOP:
      return "RUN_JS_BUNDLE_STOP";
    case RNOHMarkerId::CREATE_REACT_CONTEXT_START:
      return "CREATE_REACT_CONTEXT_START";
    case RNOHMarkerId::CREATE_REACT_CONTEXT_STOP:
      return "CREATE_REACT_CONTEXT_STOP";
    case RNOHMarkerId::JS_BUNDLE_STRING_CONVERT_START:
      return "JS_BUNDLE_STRING_CONVERT_START";
    case RNOHMarkerId::JS_BUNDLE_STRING_CONVERT_STOP:
      return "JS_BUNDLE_STRING_CONVERT_STOP";
    case RNOHMarkerId::NATIVE_MODULE_SETUP_START:
      return "NATIVE_MODULE_SETUP_START";
    case RNOHMarkerId::NATIVE_MODULE_SETUP_STOP:
      return "NATIVE_MODULE_SETUP_STOP";
    case RNOHMarkerId::REGISTER_JS_SEGMENT_START:
      return "REGISTER_JS_SEGMENT_START";
    case RNOHMarkerId::REGISTER_JS_SEGMENT_STOP:
      return "REGISTER_JS_SEGMENT_STOP";
    case RNOHMarkerId::REACT_INSTANCE_INIT_START:
      return "REACT_INSTANCE_INIT_START";
    case RNOHMarkerId::REACT_INSTANCE_INIT_STOP:
      return "REACT_INSTANCE_INIT_STOP";
    case RNOHMarkerId::CONTENT_APPEARED:
      return "CONTENT_APPEARED";
    case RNOHMarkerId::DOWNLOAD_START:
      return "DOWNLOAD_START";
    case RNOHMarkerId::DOWNLOAD_END:
      return "DOWNLOAD_END";
    case RNOHMarkerId::REACT_BRIDGE_LOADING_START:
      return "REACT_BRIDGE_LOADING_START";
    case RNOHMarkerId::REACT_BRIDGE_LOADING_END:
      return "REACT_BRIDGE_LOADING_END";
    case RNOHMarkerId::ON_HOST_RESUME_START:
      return "ON_HOST_RESUME_START";
    case RNOHMarkerId::ON_HOST_RESUME_END:
      return "ON_HOST_RESUME_END";
    case RNOHMarkerId::ON_HOST_PAUSE_START:
      return "ON_HOST_PAUSE_START";
    case RNOHMarkerId::ON_HOST_PAUSE_END:
      return "ON_HOST_PAUSE_END";
    case RNOHMarkerId::CHANGE_THREAD_PRIORITY:
      return "CHANGE_THREAD_PRIORITY";
    case RNOHMarkerId::INITIALIZE_MODULE_START:
      return "INITIALIZE_MODULE_START";
    case RNOHMarkerId::INITIALIZE_MODULE_END:
      return "INITIALIZE_MODULE_END";
    case RNOHMarkerId::PROCESS_PACKAGES_START:
      return "PROCESS_PACKAGES_START";
    case RNOHMarkerId::PROCESS_PACKAGES_END:
      return "PROCESS_PACKAGES_END";
    case RNOHMarkerId::CREATE_UI_MANAGER_MODULE_START:
      return "CREATE_UI_MANAGER_MODULE_START";
    case RNOHMarkerId::CREATE_UI_MANAGER_MODULE_END:
      return "CREATE_UI_MANAGER_MODULE_END";
    case RNOHMarkerId::CREATE_MODULE_START:
      return "CREATE_MODULE_START";
    case RNOHMarkerId::CREATE_MODULE_END:
      return "CREATE_MODULE_END";
    case RNOHMarkerId::PROCESS_CORE_REACT_PACKAGE_START:
      return "PROCESS_CORE_REACT_PACKAGE_START";
    case RNOHMarkerId::PROCESS_CORE_REACT_PACKAGE_END:
      return "PROCESS_CORE_REACT_PACKAGE_END";
    case RNOHMarkerId::FABRIC_COMMIT_START:
      return "FABRIC_COMMIT_START";
    case RNOHMarkerId::FABRIC_COMMIT_END:
      return "FABRIC_COMMIT_END";
    case RNOHMarkerId::FABRIC_FINISH_TRANSACTION_START:
      return "FABRIC_FINISH_TRANSACTION_START";
    case RNOHMarkerId::FABRIC_FINISH_TRANSACTION_END:
      return "FABRIC_FINISH_TRANSACTION_END";
    case RNOHMarkerId::FABRIC_DIFF_START:
      return "FABRIC_DIFF_START";
    case RNOHMarkerId::FABRIC_DIFF_END:
      return "FABRIC_DIFF_END";
    case RNOHMarkerId::FABRIC_LAYOUT_START:
      return "FABRIC_LAYOUT_START";
    case RNOHMarkerId::FABRIC_LAYOUT_END:
      return "FABRIC_LAYOUT_END";
    case RNOHMarkerId::FABRIC_BATCH_EXECUTION_START:
      return "FABRIC_BATCH_EXECUTION_START";
    case RNOHMarkerId::FABRIC_BATCH_EXECUTION_END:
      return "FABRIC_BATCH_EXECUTION_END";
    case RNOHMarkerId::FABRIC_UPDATE_UI_MAIN_THREAD_START:
      return "FABRIC_UPDATE_UI_MAIN_THREAD_START";
    case RNOHMarkerId::FABRIC_UPDATE_UI_MAIN_THREAD_END:
      return "FABRIC_UPDATE_UI_MAIN_THREAD_END";
    default:
      DLOG(WARNING) << "Unknown RNOHMarkerId " << static_cast<int>(markerId);
      return "UNKNOWN";
  }
}

RNOHMarker::RNOHMarkerId RNOHMarker::harmonyMarkerIdForString(
    const std::string& markerId) {
  if (markerId == "CREATE_REACT_CONTEXT_START") {
    return RNOHMarkerId::CREATE_REACT_CONTEXT_START;
  }
  if (markerId == "DOWNLOAD_START") {
    return RNOHMarkerId::DOWNLOAD_START;
  }
  if (markerId == "DOWNLOAD_END") {
    return RNOHMarkerId::DOWNLOAD_END;
  }
  throw std::invalid_argument("Unknown markerId: " + markerId);
}

} // namespace rnoh

File: packages/tester/harmony/react_native_openharmony/src/main/ets/RNOH/RNOHMarker.ts
/**
 * Copyright (c) 2024 Huawei Technologies Co., Ltd.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export interface RNOHMarkerEventPayload {
  markerId: string;
  tag: string;
  timestamp: number;
}

/**
 * @actor RNOH_LIBRARY
 * Implement this interface to register callback when RNOH markers are logged
 */
export interface RNOHMarkerListener {
  /**
   * This callback is triggered when an RNOH marker is logged
   *
   * @param markerId See RNOHMarkerId
   * @param tag The tag provided to logMarker
   * @param timestamp Only the difference between two timestamps is meaningful.
   * Unit is ms
   */
  logMarker(markerId: RNOHMarkerId, tag: string, timestamp: number): void;
}

/**
 * @actor RNOH_LIBRARY
 * Important events happening throughout an RN app's lifecycle
 * @note to framework developers. Keep this enum in sync with the one defined
 * in RNOHMarker.h
 */
export enum RNOHMarkerId {
  UNKNOWN_REACT_MARKER,
  APP_STARTUP_START,
  APP_STARTUP_STOP,
  INIT_REACT_RUNTIME_START,
  INIT_REACT_RUNTIME_STOP,
  NATIVE_REQUIRE_START,
  NATIVE_REQUIRE_STOP,
  RUN_JS_BUNDLE_START,
  RUN_JS_BUNDLE_STOP,
  CREATE_REACT_CONTEXT_START,
  CREATE_REACT_CONTEXT_STOP,
  JS_BUNDLE_STRING_CONVERT_START,
  JS_BUNDLE_STRING_CONVERT_STOP,
  NATIVE_MODULE_SETUP_START,
  NATIVE_MODULE_SETUP_STOP,
  REGISTER_JS_SEGMENT_START,
  REGISTER_JS_SEGMENT_STOP,
  REACT_INSTANCE_INIT_START,
  REACT_INSTANCE_INIT_STOP,
  CONTENT_APPEARED,
  DOWNLOAD_START,
  DOWNLOAD_END,
  REACT_BRIDGE_LOADING_START,
  REACT_BRIDGE_LOADING_END,
  ON_HOST_RESUME_START,
  ON_HOST_RESUME_END,
  ON_HOST_PAUSE_START,
  ON_HOST_PAUSE_END,
  CHANGE_THREAD_PRIORITY,
  INITIALIZE_MODULE_START,
  INITIALIZE_MODULE_END,
  PROCESS_PACKAGES_START,
  PROCESS_PACKAGES_END,
  CREATE_UI_MANAGER_MODULE_START,
  CREATE_UI_MANAGER_MODULE_END,
  CREATE_MODULE_START,
  CREATE_MODULE_END,
  PROCESS_CORE_REACT_PACKAGE_START,
  PROCESS_CORE_REACT_PACKAGE_END,
  FABRIC_COMMIT_START,
  FABRIC_COMMIT_END,
  FABRIC_FINISH_TRANSACTION_START,
  FABRIC_FINISH_TRANSACTION_END,
  FABRIC_DIFF_START,
  FABRIC_DIFF_END,
  FABRIC_LAYOUT_START,
  FABRIC_LAYOUT_END,
  FABRIC_BATCH_EXECUTION_START,
  FABRIC_BATCH_EXECUTION_END,
  FABRIC_UPDATE_UI_MAIN_THREAD_START,
  FABRIC_UPDATE_UI_MAIN_THREAD_END,
}

/**
 * @actor RNOH_LIBRARY
 * Registers listeners to receive RNOH marker events
 */
export class RNOHMarker {
  private static listeners: Set<RNOHMarkerListener> = new Set();

  /**
   * Registers a listener to receive RNOH marker events
   * @param listener
   */
  static addListener(listener: RNOHMarkerListener) {
    this.listeners.add(listener);
  }

  /**
   * Unregisters a listener to stop receiving RNOH marker events
   * @param listener
   */
  static removeListener(listener: RNOHMarkerListener) {
    this.listeners.delete(listener);
  }

  /**
   * @internal
   */
  static notifyListeners(markerId: string, tag: string, timestamp: number) {
    const resolvedMarkerId =
      RNOHMarkerId[markerId as keyof typeof RNOHMarkerId] ||
      RNOHMarkerId.UNKNOWN_REACT_MARKER;
    this.listeners.forEach(listener =>
      listener.logMarker(resolvedMarkerId, tag, timestamp),
    );
  }
}
